metaflow/metaflow_git.py

```diff
@@ -0,0 +1,131 @@
+                    domain = parts[0].replace("git@", "")
+                    repo_path = parts[1]
+                    url = f"https://{domain}/{repo_path}"
+            return url
+        return None
+    except (OSError, subprocess.SubprocessError):
+        return None
+
+
+def _get_branch_name(path: Union[str, os.PathLike]):
+    """Get the current git branch name"""
+    try:
+        result = subprocess.run(
+            [GIT_COMMAND, "rev-parse", "--abbrev-ref", "HEAD"],
+            cwd=path,
+            capture_output=True,
+            text=True,
+            check=False,
+        )
+        return result.stdout.strip() if result.returncode == 0 else None
+    except (OSError, subprocess.SubprocessError):
+        return None
+
+
+def _get_commit_sha(path: Union[str, os.PathLike]):
+    """Get the current git commit SHA"""
+    try:
+        result = subprocess.run(
+            [GIT_COMMAND, "rev-parse", "HEAD"],
+            cwd=path,
+            capture_output=True,
+            text=True,
+            check=False,
+        )
+        return result.stdout.strip() if result.returncode == 0 else None
+    except (OSError, subprocess.SubprocessError):
+        return None
+
+
+def _is_in_git_repo(path: Union[str, os.PathLike]):
+    """Check if we're currently in a git repository"""
+    try:
+        result = subprocess.run(
+            [GIT_COMMAND, "rev-parse", "--is-inside-work-tree"],
+            cwd=path,
+            capture_output=True,
+            text=True,
+            check=False,
+        )
+        return result.returncode == 0 and result.stdout.strip() == "true"
+    except (OSError, subprocess.SubprocessError):
+        return False
+
+
+def _has_uncommitted_changes(path: Union[str, os.PathLike]):
+    """Check if the git repository has uncommitted changes"""
+    try:
+        result = subprocess.run(
+            [GIT_COMMAND, "status", "--porcelain"],
+            cwd=path,
+            capture_output=True,
+            text=True,
+            check=False,
+        )
+        # If output is not empty, there are uncommitted changes
+        return result.returncode == 0 and bool(result.stdout.strip())
+    except (OSError, subprocess.SubprocessError):
+        return None
+
+
+def get_repository_info(path: Union[str, os.PathLike]) -> Dict[str, Union[str, bool]]:
+    """Get git repository information for a path
+
+    Returns:
+        dict: Dictionary containing:
+            repo_url: Repository URL (converted to HTTPS if from SSH)
+            branch_name: Current branch name
+            commit_sha: Current commit SHA
+            has_uncommitted_changes: Boolean indicating if there are uncommitted changes
+    """
+    global _git_info_cache
+
+    if _git_info_cache is not None:
+        return _git_info_cache
+
+    _git_info_cache = {}
+    if _is_in_git_repo(path):
+        _git_info_cache = {
+            "repo_url": _get_repo_url(path),
+            "branch_name": _get_branch_name(path),
```

Comment:
I was thinking something like this:
```
echo "$(git rev-parse --abbrev-ref HEAD) $(git rev-parse --short HEAD)$(git diff --quiet || echo ' dirty')"
```
(one subprocess and also one single metadata element -- it doesn't save on the multitude of invocation of git but does only launch a single subprocesse, well two after we check that git is in the right directory but we could also probably skip that because the git command will error out if not in a git repo)

Reply:
+1

Reply:
it just unnecessarily complicates the code without any real performance benefit

Reply:
If you want separate calls sure (although there is a performance benefit but I agree it's tiny) -- not sure what it complicates but that's more stylistic. For the fewer fields of metadata, I would *definitely* lean in that direction. Metadata calls are *not* cheap (as you previously pointed out) and adding 4 fields seems a bit much.

Reply:
we can lob everything into a json value to reduce the number of rows - @saikonen can you look into that too?